### PR TITLE
dockershim hostport manager should use HostIP

### PR DESCRIPTION
**Pull request by aojea**
/kind bug

**What this PR does / why we need it**:

the hostport manager was not taking into consideration the hostIP
when binding the socket of the hostPort, causing that the same
HostPort can not be used with different IP addresses.

```release-note
fix a bug where the hostport manager used by dockershim didn't allow to bind the same HostPort on different HostIPs
```

### CODE REVIEW

**Comment by aojea** [comment]
/assign @thockin @danwinship @dcbw 

I need to check if there is some job exercising this code with e2e tests, lot of jobs were move to containerd and (I think) using the CNI portmap plugin (at least the KIND ones)

**Comment by thockin** [comment]
I am frankly surprised this works at all in the scheduler - does it?  Do we really want it to?  What's the use case?  This seems verrrrry subtle.

**Comment by aojea** [comment]
> I am frankly surprised this works at all in the scheduler - does it? Do we really want it to? What's the use case? This seems verrrrry subtle.

There is a conformance test covering exactly this use case
https://github.com/kubernetes/kubernetes/blob/62b8faed4042e9f1a9c1e895ad1c36bb473eb6ee/test/e2e/scheduling/predicates.go#L654-L660

It seems that works (at least in CRIO https://github.com/cri-o/cri-o/pull/4130#issue-475492178) because, after the first attempt to allocate the HostPort, it removes ALL sockets listening in that port, so the second attempt works

**Comment by aojea** [comment]
I can't find any job in the CI running this code :/

I was reading the issue tracking the hostport implementation #35457, and it seems that this was temporary until the CNI plugins implements the portmapping,

@thockin  do you think that does not make sense to pursue this and let it as is, fact is nobody noticed this based on a quick github search.

If we don fix it here., @danwinship I can make it a package in crio instead of vendoring it

**Comment by danwinship** [comment]
> do you think that does not make sense to pursue this and let it as is, fact is nobody noticed this based on a quick github search.

Well, besides the fact that we know some people are vendoring this code, there's also the fact that it means that people using kubernetes with dockershim get non-conforming behavior. The fact that we don't currently run any CI jobs in that configuration doesn't change the fact that it's broken.

The fact that the conformance test fails to accurately test the behavior it's supposed to be testing is... interesting. We should probably fix the conformance test in 1.20 (by making it check that the first hostport still connects to the first pod after the second pod has been created) but not backport the fix, so as to not retroactively de-conform systems that previously passed conformance in earlier releases. At least, I think that's the right approach?

File: pkg/kubelet/dockershim/network/hostport/hostport_manager_test.go [hunk 1]
```diff
@@ -106,13 +136,18 @@ func TestOpenCloseHostports(t *testing.T) {
 		execer:      exec.New(),
 	}
 
+	// open all hostports defined in the test cases
 	for _, tc := range openPortCases {
 		mapping, err := manager.openHostports(tc.podPortMapping)
+		for hostport, socket := range mapping {
+			manager.hostPortMap[hostport] = socket
```
Comment (danwinship) [inline pkg/kubelet/dockershim/network/hostport/hostport_manager_test.go:143]: Why is this needed here? This seems like something that should either happen in the actual manager code, or else not be needed.

If it's just for the counting below, then that's really only testing the behavior of the test program, not the behavior of the manager...

File: pkg/kubelet/dockershim/network/hostport/hostport_manager_test.go [hunk 2]
```diff
@@ -372,6 +510,8 @@ func TestHostportManager(t *testing.T) {
 	for _, port := range portOpener.mem {
 		assert.EqualValues(t, true, port.closed)
 	}
+	// Clear all elements in hostPortMap
+	assert.Zero(t, len(manager.hostPortMap))
```
Comment (danwinship) [inline pkg/kubelet/dockershim/network/hostport/hostport_manager_test.go:514]: that's not clearing the elements, it's asserting that someone else already did

Comment (aojea) [reply pkg/kubelet/dockershim/network/hostport/hostport_manager_test.go:514]: yeah, I wonder what I was thinking :)

File: pkg/kubelet/dockershim/network/hostport/hostport_manager_test.go (hunk 1 above)
Comment (aojea) [reply pkg/kubelet/dockershim/network/hostport/hostport_manager_test.go:143]: this was actually a bug in the test.
The same loop to add the open sockets to the hostPortMap map does happen in the manager.Add() method, https://github.com/kubernetes/kubernetes/blob/12fc5e50b9394cc3fa498b24a96c894d96b3bff2/pkg/kubelet/dockershim/network/hostport/hostport_manager.go#L107-L114

We need this because later we test the `closeHostports()` method that iterates over that map https://github.com/kubernetes/kubernetes/blob/12fc5e50b9394cc3fa498b24a96c894d96b3bff2/pkg/kubelet/dockershim/network/hostport/hostport_manager.go#L310-L327
in 
```
	for _, tc := range closePortCases {
		err := manager.closeHostports(tc.portMappings)
```

if we don't populate the map the closeHostports will always succeed because it's map. 
I've also added the counting below as a safety check and to make it easier to debug

File: pkg/kubelet/dockershim/network/hostport/hostport_manager_test.go [hunk 3]
```diff
@@ -35,13 +35,15 @@ func TestOpenCloseHostports(t *testing.T) {
 		podPortMapping *PodPortMapping
 		expectError    bool
 	}{
+		// no portmaps
```
[the remark this answers is not in the document]
Comment (aojea) [reply pkg/kubelet/dockershim/network/hostport/hostport_manager_test.go:38]: Naming is not useful here because the test cases are not independent. Personally, I don't like using this test structure but didn't want to change the whole test,so I carried over with current status. At least the comments explain better what is supposed to happen

**Comment by thockin** [comment]
I did some quick tests on my own cluster and, shock, it does seem to work.
I dislike this very much.  It seems like a great opportunity to make a
mistake (honestly, hostIP at all seems poorly thought out but that dates to
2014, sigh).

So the choice is really a) fix it in dockershim to match containerd or b)
don't allow same-port-different-IP between pods.

I want to say B, but I suspect the answer is A.  Can we document this
behavior on the API types.go, please?

File: pkg/kubelet/dockershim/network/hostport/hostport_manager.go [hunk 4]
```diff
@@ -152,10 +152,17 @@ func (hm *hostportManager) Add(id string, podPortMapping *PodPortMapping, natInt
 
 		// DNAT to the podIP:containerPort
 		hostPortBinding := net.JoinHostPort(podIP, strconv.Itoa(int(pm.ContainerPort)))
-		writeLine(natRules, "-A", string(chain),
-			"-m", "comment", "--comment", fmt.Sprintf(`"%s hostport %d"`, podFullName, pm.HostPort),
-			"-m", protocol, "-p", protocol,
-			"-j", "DNAT", fmt.Sprintf("--to-destination=%s", hostPortBinding))
+		if pm.HostIP == "" || pm.HostIP == "0.0.0.0" || pm.HostIP == "::" {
```
Comment (andrewsykim) [inline pkg/kubelet/dockershim/network/hostport/hostport_manager.go:155]: Does kubelet know to use iptables for IPv6 if HostIP is `::` here?

Comment (aojea) [reply pkg/kubelet/dockershim/network/hostport/hostport_manager.go:155]: kubelet is ALWAYS single stack as today, and its IP family depends on the nodeIP, 

https://github.com/kubernetes/kubernetes/blob/c236285708d43ad7535a37cbff41fa380afc199d/pkg/kubelet/kubelet.go#L468-L473

The hostport_manager has its own iptables handler, but depends on the kubelet for some rules, however, due to this https://github.com/kubernetes/kubernetes/issues/82125 , those rules are added by kube-proxy

Now that you mention it, without kube-proxy this won't work for dual-stack @danwinship

**Comment by aojea** [comment]
> I want to say B, but I suspect the answer is A. Can we document this behavior on the API types.go, please?

I guess it is  https://github.com/kubernetes/kubernetes/blob/adac1bfb8b52def074022e0a3b07740fdc1a7adf/pkg/apis/core/types.go#L1762-L1764

File: pkg/kubelet/dockershim/network/hostport/hostport_manager_test.go [hunk 5]
```diff
@@ -319,24 +439,41 @@ func TestHostportManager(t *testing.T) {
...
+		`:KUBE-HP-TUKTZ736U5JD5UTK - [0:0]`: true,
+		`:KUBE-HP-CAAJ45HDITK7ARGM - [0:0]`: true,
+		`:KUBE-HP-WFUNFVXVDLD5ZVXN - [0:0]`: true,
+		`:KUBE-HP-4MFWH2F2NAOMYD6A - [0:0]`: true,
+		"-A KUBE-HOSTPORTS -m comment --comment \"pod3_ns1 hostport 8443\" -m tcp -p tcp --dport 8443 -j KUBE-HP-WFBOALXEP42XEMJK":                        true,
+		"-A KUBE-HOSTPORTS -m comment --comment \"pod1_ns1 hostport 8081\" -m udp -p udp --dport 8081 -j KUBE-HP-63UPIDJXVRSZGSUZ":                        true,
+		"-A KUBE-HOSTPORTS -m comment --comment \"pod1_ns1 hostport 8080\" -m tcp -p tcp --dport 8080 -j KUBE-HP-IJHALPHTORMHHPPK":                        true,
+		"-A KUBE-HOSTPORTS -m comment --comment \"pod1_ns1 hostport 8083\" -m sctp -p sctp --dport 8083 -j KUBE-HP-XU6AWMMJYOZOFTFZ":                      true,
+		"-A KUBE-HOSTPORTS -m comment --comment \"pod5_ns5 hostport 8888\" -m tcp -p tcp --dport 8888 -j KUBE-HP-TUKTZ736U5JD5UTK":                        true,
+		"-A KUBE-HOSTPORTS -m comment --comment \"pod5_ns5 hostport 8888\" -m tcp -p tcp --dport 8888 -j KUBE-HP-CAAJ45HDITK7ARGM":                        true,
```
Comment (aojea) [inline pkg/kubelet/dockershim/network/hostport/hostport_manager_test.go:451]: actually this does not seem to work as expected, it should load balance between the different backend HostIPs

**Comment by aojea** [comment]
I just realized that the hostport_ syncer code was removed recently 
https://github.com/kubernetes/kubernetes/pull/93581

In the kubernetes repo only the kubenet seems to use this code :thinking:

**Comment by SergeyKanzhelev** [comment]
> I just realized that the hostport_ syncer code was removed recently
> #93581
> 
> In the kubernetes repo only the kubenet seems to use this code 🤔

My PR was just a follow up clean up after this one: https://github.com/kubernetes/kubernetes/pull/62212 I wasn't sure about hostport functionality and whether it is actually used.

One more discussion point here - there is a leftover of duplicates suppression logic for [nodeport mappings](https://github.com/kubernetes/kubernetes/blob/19064f47389429f8360e1d6684bef3d770db9368/pkg/kubelet/container/helpers.go#L324-L345). Is it needed in kubelet if CNI should be responsible for setting it all up?

**Comment by aojea** [comment]
> One more discussion point here - there is a leftover of duplicates suppression logic for [nodeport mappings](https://github.com/kubernetes/kubernetes/blob/19064f47389429f8360e1d6684bef3d770db9368/pkg/kubelet/container/helpers.go#L324-L345). Is it needed in kubelet if CNI should be responsible for setting it all up?

@thockin I think that this is only being used by kubenet in k8s, the rest of the imports are just for the type `[]*hostport.PortMapping`

is kubenet going to move out of the kubelet?
